src/declare.rs

```diff
@@ -237,6 +244,87 @@ impl Drop for ClassDecl {
     }
 }
 
+/// Describes a method of an Objective-C protocol.
+pub trait MethodDescription {
+    /// The arguments to the protocol method.
+    type Args: EncodeArguments;
+
+    /// The return type of the protocol method.
+    type Ret: Encode;
+}
+
+impl<Args: EncodeArguments, Ret: Encode> MethodDescription for (Args, Ret) {
+    type Args = Args;
+    type Ret = Ret;
+}
+
+/// A type for declaring a new protocol and adding new methods to it
+/// before registering it.
+pub struct ProtocolDecl {
+    proto: *mut Protocol
+}
+
+impl ProtocolDecl {
+    /// Constructs a `ProtocolDecl` with the given name. Returns `None` if the
+    /// protocol couldn't be allocated.
+    pub fn new(name: &str) -> Option<ProtocolDecl> {
+        let c_name = CString::new(name).unwrap();
+        let proto = unsafe {
+            runtime::objc_allocateProtocol(c_name.as_ptr())
+        };
+        if proto.is_null() {
+            None
+        } else {
+            Some(ProtocolDecl { proto: proto })
+        }
+    }
+
+    fn add_method_description_common<M: MethodDescription>(&mut self, sel: Sel, is_required: bool,
+            is_instance_method: bool) {
+        let encs = M::Args::encodings();
+        let encs = encs.as_ref();
+        let sel_args = count_args(sel);
+        assert!(sel_args == encs.len(),
+            "Selector accepts {} arguments, but function accepts {}",
+            sel_args, encs.len(),
+        );
+        let types = method_type_encoding(&M::Ret::encode(), encs);
+        unsafe {
+            runtime::protocol_addMethodDescription(
+                self.proto, sel, types.as_ptr(), is_required as BOOL, is_instance_method as BOOL);
+        }
+    }
+
+    /// Adds an instance method declaration with a given description to self.
+    pub fn add_method_description<M: MethodDescription>(&mut self, sel: Sel, is_required: bool) {
+        self.add_method_description_common::<M>(sel, is_required, true)
+    }
+
+    /// Adds a class method declaration with a given description to self.
+    pub fn add_class_method_description<M: MethodDescription>(&mut self, sel: Sel, is_required: bool) {
+        self.add_method_description_common::<M>(sel, is_required, false)
+    }
+
+    /// Adds a requirement on another protocol.
+    pub fn add_protocol(&mut self, proto: &Protocol) {
+        let proto_ptr: *const _ = &*proto;
+        unsafe {
+            runtime::protocol_addProtocol(self.proto, proto_ptr);
+        }
+    }
+
+    /// Registers self, consuming it and returning a reference to the
+    /// newly registered `Protocol`.
+    pub fn register(self) -> &'static Protocol {
+        unsafe {
+            let proto = self.proto;
+            runtime::objc_registerProtocol(proto);
+            mem::forget(self);
```

Comment:
`mem::forget` here prevents a destructor from being run but it doesn't look like `ProtocolDecl` implements `Drop` :)